src/m365/tenant/commands/auditlog/auditlog-report.ts

```diff
@@ -0,0 +1,246 @@
+          const requestOptions: any = {
+            url: `${this.serviceUrl}/${this.tenantId}/${startSubscriptionEndPoint}`,
+            headers: {
+              accept: 'application/json;odata.metadata=none'
+            },
+            responseType: 'json'
+          };
+
+          return request.post(requestOptions);
+        }
+
+        return Promise.resolve();
+      });
+  }
+
+  private getAuditContentList(args: CommandArgs, logger: Logger): Promise<AuditContentList[]> {
+    if (this.verbose) {
+      logger.logToStderr(`Start listing Audit Content URL`);
+    }
+
+    let SubscriptionListEndpoint: string = (typeof args.options.startTime !== 'undefined' && typeof args.options.endTime !== 'undefined') ?
+      `/activity/feed/subscriptions/content?contentType=${(<any>AuditContentTypes)[args.options.contentType]}&PublisherIdentifier=${this.tenantId}&starttime=${args.options.startTime}&endTime=${args.options.endTime}` :
+      `/activity/feed/subscriptions/content?contentType=${(<any>AuditContentTypes)[args.options.contentType]}&PublisherIdentifier=${this.tenantId}`;
+    const requestOptions: any = {
+      url: `${this.serviceUrl}/${this.tenantId}/${SubscriptionListEndpoint}`,
+      headers: {
+        accept: 'application/json;odata.metadata=none'
+      },
+      responseType: 'json'
+    };
+
+    return request.get<AuditContentList[]>(requestOptions)
+  }
+
+  private async getBatchedAuditLogReports(logger: Logger, AuditContentLists: AuditContentList[]) : Promise<any> {
+    if (this.verbose) {
+      logger.logToStderr(`Start generating Audit Reports in batchwise manner`);
+    }
+    
+    //Batching Approach - Given Batch size is 10
+    const batchSize: number = 10;
+    let runningBatch: number = 0;
+    var CompleteAuditReportsUnflattened : any = [];
+    //Temporary - Will be removed in production
+    let tempMaximumCount: number = 20;
+
+    for (let i = 0; i < (AuditContentLists.length<tempMaximumCount?AuditContentLists.length:tempMaximumCount); i += batchSize) {
+      if (this.verbose) {
+        runningBatch++;
+        logger.logToStderr(`Generating Audit Reports for Batch : ${runningBatch}`);
+      }
+
+      const PromiseRequestBatch = AuditContentLists.slice(i, i + batchSize<AuditContentLists.length?i+batchSize:AuditContentLists.length).map((AuditContentList) => {
+        //logger.log(`Inner Loop : ${i}`);
+        return this.getAuditLogReportforSingleContentURL(AuditContentList.contentUri);
+      })
+      
+      //Async/AWAIT Approach
+      CompleteAuditReportsUnflattened.push(await Promise.all(PromiseRequestBatch));
+
+      // //NOT Async/AWAIT Approach
+      // return Promise.all(PromiseRequestBatch)
+      // .then((batchedAuditReport): void => {
+      //   logger.log(`${batchedAuditReport[0][0]}`)
+      //   CompleteAuditReportsUnflattened.push(batchedAuditReport);
+      // });
```

Comment:
Like mentioned in the above comment, I am unable to Chain multiple batches of promises to get a result. Can you please advise what could be done here.

Reply:
Thanks for reaching out @arjunumenon, I think we'd need something similar to recursion we use elsewhere in the CLI where we pass the promise and resolve only after the last operation was completed.
Here is some pseudo-code that could likely help:

```ts
class YourCommand {
  // array with sets of requests to execute; need to define here so that we can access it from subsequent calls
  private requestsChunks;

  public commandAction(args, cb) {
    // build all requests here
    // split requests into chunks of 10
    this.requestsChunks = MyCommand.chunkArray(allRequests, 10);
    new Promise((resolve, reject) => {
         this.getData(0, resolve, reject);
      })
      .then(_ => {
         // all requests done; log output and close command
         cb();
      })
  }

  private getData(chunkNumber, resolve, reject) {
    Promise
      .all(this.requestsChunks[chunkNumber])
      .then(data => {
        // store retrieved data
        // see if there are chunks left that we should execute; if not, resolve the root promise
        if (chunkNumber < this.requestsChunks.length-1) {
          this.getData(++chunkNumber, resolve, reject)
        }
        else {
          resolve();
        }
     }, err => reject(err));
  }

  // from: https://ourcodeworld.com/articles/read/278/how-to-split-an-array-into-chunks-of-the-same-size-easily-in-javascript
  private static chunkArray(myArray, chunkSize){
    var results = [];
    
    while (myArray.length) {
        results.push(myArray.splice(0, chunkSize));
    }
    
    return results;
  }
}
```

Reply:
Thanks a lot @waldekmastykarz . That is why they say, all you need is a brilliant third eye and your clouded path will be clear like a sunshine.. 
Thanks a lot for spending time and sending the snippet. I had tried recursion in a different way, but did not yield the expected result. But the approach you have shared seems to be promising with all **Promises** aligned in a better way.🚀

Will try that out and hopefully would be completed without bothering you again.. Thanks again. 🙏

Reply:
I'm getting stuck at times too and having someone else look at the problem often helps. Never hesitate to reach out! We're in this together 👏